[motion-path] Accept position parameter in ray()

nmoucht opened the pull request:
#### f665fe5b944c75b267e7bb6900d096d99e4e45d5
<pre>
[motion-path] Accept position parameter in ray()
<a href="https://bugs.webkit.org/show_bug.cgi?id=258113">https://bugs.webkit.org/show_bug.cgi?id=258113</a>
rdar://110821024

Reviewed by Tim Nguyen.

WIP from Tim Nguyen.

The ray() function should accept an at &lt;position&gt; parameter which represents the starting position of the ray.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/motion/animation/offset-path-interpolation-005-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/motion/offset-path-ray-013.html:
* LayoutTests/imported/w3c/web-platform-tests/css/motion/parsing/offset-path-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/motion/parsing/offset-path-parsing-valid-expected.txt:
* Source/WebCore/css/CSSRayValue.cpp:
(WebCore::CSSRayValue::customCSSText const):
(WebCore::CSSRayValue::equals const):
* Source/WebCore/css/CSSRayValue.h:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::valueForPathOperation):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeRayShape):
* Source/WebCore/rendering/PathOperation.cpp:
(WebCore::RayPathOperation::create):
(WebCore::RayPathOperation::clone const):
(WebCore::RayPathOperation::blend const):
(WebCore::RayPathOperation::currentOffset const):
(WebCore::RayPathOperation::lengthForPath const):
(WebCore::RayPathOperation::getPath const):
* Source/WebCore/rendering/PathOperation.h:
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::offsetFromContainer):
(WebCore::RenderLayer::setReferenceBoxForPathOperations):
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertPathOperation):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/267459@main">https://commits.webkit.org/267459@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4ff4df6758b9dfbca3ff47888e11fccae62b1ffd

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/16574 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/16897 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/17339 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/18356 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/15544 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/16764 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/20127 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/17038 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/17873 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/16772 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/17174 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/14339 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/19139 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/14418 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/15023 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/21809 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/15406 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/15189 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/19499 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/15787 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/13405 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/14985 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/19354 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2051 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/15613 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/css/CSSRayValue.h:
```diff
@@ -36,32 +36,38 @@ namespace WebCore {
 class CSSRayValue final : public CSSValue {
 public:
-    static Ref<CSSRayValue> create(Ref<CSSPrimitiveValue>&& angle, CSSValueID size, bool isContaining)
+    static Ref<CSSRayValue> create(Ref<CSSPrimitiveValue>&& angle, CSSValueID size, bool isContaining, RefPtr<CSSValue>&& positionX, RefPtr<CSSValue>&& positionY)
```
I agree with @rreno's feedback from the other PR that we should just take and store the position as a CSSValuePair on CSSRayValue instead of 2 separate members.

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/style/StyleBuilderConverter.h:
```diff
@@ -751,6 +751,12 @@ inline RefPtr<PathOperation> BuilderConverter::convertPathOperation(BuilderState
 
+        bool hasPosition = rayValue.positionX();
+        if (hasPosition) {
+            LengthPoint position { convertPositionComponentX(builderState, *rayValue.positionX()), convertPositionComponentY(builderState, *rayValue.positionY()) };
```
...and convertPositionComponent can convert CSSValuePair to LengthPoint.

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/css/CSSRayValue.cpp:
```diff
@@ -34,12 +34,17 @@ namespace WebCore {
+        hasPosition ? " at "_s : ""_s,
+        hasPosition ? m_position->first().cssText() : ""_s,
+        hasPosition ? " "_s : ""_s,
+        hasPosition ? m_position->second().cssText() : ""_s,
```
```suggestion
        hasPosition ? m_position->cssText() : ""_s,
```
I think you can just do this and get the right serialization, CSSValuePair has a customCSSText() that should work properly here.

nmoucht replied:
I tried that but it only gave the serialization for the first value for whatever reason.

nmoucht replied:
Ah I see need to set IdenticalValueSerialization.

nmoucht replied:
Fixed.

nmoucht committed:
[motion-path] Accept position parameter in ray()
https://bugs.webkit.org/show_bug.cgi?id=258113
rdar://110821024

Reviewed by Tim Nguyen.

WIP from Tim Nguyen.

The ray() function should accept an at <position> parameter which represents the starting position of the ray.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/motion/animation/offset-path-interpolation-005-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/motion/offset-path-ray-013.html:
* LayoutTests/imported/w3c/web-platform-tests/css/motion/parsing/offset-path-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/motion/parsing/offset-path-parsing-valid-expected.txt:
* Source/WebCore/css/CSSRayValue.cpp:
(WebCore::CSSRayValue::customCSSText const):
(WebCore::CSSRayValue::equals const):
* Source/WebCore/css/CSSRayValue.h:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::valueForPathOperation):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeRayShape):
* Source/WebCore/rendering/PathOperation.cpp:
(WebCore::RayPathOperation::create):
(WebCore::RayPathOperation::clone const):
(WebCore::RayPathOperation::blend const):
(WebCore::RayPathOperation::currentOffset const):
(WebCore::RayPathOperation::lengthForPath const):
(WebCore::RayPathOperation::getPath const):
* Source/WebCore/rendering/PathOperation.h:
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::offsetFromContainer):
(WebCore::RenderLayer::setReferenceBoxForPathOperations):
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertPathOperation):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/267459@main
